File: src/main/java/com/janrain/backplane/server2/BackplaneMessage.java
/*
 * Copyright 2012 Janrain, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.janrain.backplane.server2;

import com.janrain.backplane.common.BackplaneServerException;
import com.janrain.backplane.common.DateTimeUtils;
import com.janrain.backplane.common.ExternalizableCore;
import com.janrain.backplane.common.RandomUtils;
import com.janrain.backplane.servlet.InvalidRequestException;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;
import com.janrain.commons.util.Pair;
import org.apache.commons.lang.StringUtils;
import org.apache.log4j.Logger;
import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;
import java.text.ParseException;
import java.util.*;

import static com.janrain.backplane.server2.Scope.ScopeType.*;

/**
 * @author Johnny Bufu
 */
public class BackplaneMessage extends ExternalizableCore {

    // - PUBLIC

    @SuppressWarnings("UnusedDeclaration")
    public BackplaneMessage() {}

    public BackplaneMessage(String clientSourceUrl, int defaultExpireSeconds, int maxExpireSeconds, Map<String, Object> data) throws BackplaneServerException, MessageException {
        checkUpstreamExtraFields(data);
        Map<String,String> d = new LinkedHashMap<String, String>(toStringMap(data));
        String id = generateMessageId(new Date());
        d.put(Field.ID.getFieldName(), id);
        d.put(Field.SOURCE.getFieldName(), clientSourceUrl);
        if (data.containsKey(Field.PAYLOAD.getFieldName())) {
            d.put(Field.PAYLOAD.getFieldName(), extractFieldValueAsJsonString(Field.PAYLOAD, data));
        }
        Object sticky = data.get(Field.STICKY.getFieldName());
        d.put(Field.STICKY.getFieldName(), sticky != null ? sticky.toString() : Boolean.FALSE.toString());
        d.put(Field.EXPIRE.getFieldName(), DateTimeUtils.processExpireTime(sticky, data.get(Field.EXPIRE.getFieldName()), defaultExpireSeconds, maxExpireSeconds));
        super.init(id, d);
    }

    @Override
    public String getIdValue() {
        return get(Field.ID);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(Field.class);
    }

    public Pair<String, Date> updateId(Pair<String, Date> lastIdAndDate) {
        long thisTime = 0;
        try {
            thisTime = getDateFromId(getIdValue()).getTime();
        } catch (Exception e) {
            Date thisDate = new Date();
            thisTime = thisDate.getTime();
            put(Field.ID.getFieldName(), generateMessageId(thisDate));
        }

        long lastTime = 0;
        try {
            lastTime = getDateFromId(lastIdAndDate.getLeft()).getTime();
        } catch (Exception e) {
            //
        }

        if (thisTime <= lastTime) {
            logger.warn("message has an id " + getIdValue() + " that is not > the latest id of " + lastIdAndDate.getLeft());
            Date newDate = new Date(lastIdAndDate.getRight().getTime() + 1);
            put(Field.ID.getFieldName(), generateMessageId(newDate));
            logger.warn("fixed");
            return new Pair<String, Date>(getIdValue(), newDate);
        } else {
            return lastIdAndDate;
        }
    }

    public String getBus() {
        return this.get(Field.BUS);
    }

    public String getChannel() {
        return this.get(Field.CHANNEL);
    }

    public boolean isSticky() {
        return "true".equalsIgnoreCase(get(Field.STICKY));
    }

    public Map<String, Object> asFrame(String serverDomain, boolean includePayload) throws BackplaneServerException {

        HashMap<String, Object> frame = new LinkedHashMap<String, Object>();

        String serverUrl = "https://" + serverDomain + "/v2/message";
        frame.put(Field.MESSAGE_URL.getFieldName(), serverUrl + "/" + get(Field.ID));
        frame.put(Field.SOURCE.getFieldName(), get(Field.SOURCE));
        frame.put(Field.TYPE.getFieldName(), get(Field.TYPE));
        frame.put(Field.BUS.getFieldName(), get(Field.BUS));
        frame.put(Field.CHANNEL.getFieldName(), get(Field.CHANNEL));
        frame.put(Field.STICKY.getFieldName(), containsKey(Field.STICKY.getFieldName()) ? get(Field.STICKY) : Boolean.FALSE.toString());
        frame.put(Field.EXPIRE.getFieldName(), get(Field.EXPIRE));

        try {
            if (includePayload) {
                frame.put(Field.PAYLOAD.getFieldName(), (new ObjectMapper()).readValue(get(Field.PAYLOAD), Object.class) ); // un-quote the value
            }
        } catch (IOException e) {
            String errMsg = "Error deserializing message payload: " + e.getMessage();
            logger.error(errMsg);
            throw new BackplaneServerException(errMsg, e);
        }

        return frame;
    }

    public static enum Field implements MessageField {

        ID("id", NONE),

        CHANNEL("channel", FILTER),

        BUS("bus", AUTHZ_REQ),

        STICKY("sticky", false, FILTER) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                if (value != null && ! Boolean.TRUE.toString().equalsIgnoreCase(value) && ! Boolean.FALSE.toString().equalsIgnoreCase(value)) {
                    throw new InvalidRequestException("Invalid boolean value for " + getFieldName() + ": " + value);
                }
            }},

        EXPIRE("expire", true, NONE) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                try {
                    DateTimeUtils.INTERNETDATE.get().parse(value);
                } catch (ParseException e) {
                    throw new InvalidRequestException("Invalid Internet Date/Time value for " + getFieldName() + ": " + value);
                }
            }},

        SOURCE("source", FILTER) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                validateUrl(getFieldName(), value);
            }},

        TYPE("type", FILTER),

        MESSAGE_URL("messageURL", false, FILTER) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                if (StringUtils.isNotEmpty(value)) {
                    validateUrl(getFieldName(), value);
                }
            }
        },

        PAYLOAD("payload", NONE);

        @Override
        public String getFieldName() {
            return fieldName;
        }

        @Override
        public boolean isRequired() {
            return required;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(getFieldName(), value);
        }

        public Scope.ScopeType getScopeType() {
            return scopeType;
        }

        // - PRIVATE

        private String fieldName;
        private final boolean required;
        private final Scope.ScopeType scopeType;

        private Field(String fieldName, Scope.ScopeType scopeType) {
            this(fieldName, true, scopeType);
        }

        private Field(String fieldName, boolean required, Scope.ScopeType scopeType) {
            this.fieldName = fieldName;
            this.required = required;
            this.scopeType = scopeType;
        }
    }

    public static Date getDateFromId(String backplaneMessageId) {
        if (backplaneMessageId == null) {
            return null;
        }

        try {
            return DateTimeUtils.ISO8601.get().parse(backplaneMessageId.substring(0, backplaneMessageId.indexOf("Z") + 1));
        } catch (ParseException e) {
            logger.warn(e);
        }
        return null;
    }

    // - PRIVATE

    private static final long serialVersionUID = -6609794896611874473L;

    private static final Logger logger = Logger.getLogger(BackplaneMessage.class);

    private static final Set<String> UPSTREAM_FIELDS = new HashSet<String>() {{
        add(Field.TYPE.getFieldName());
        add(Field.BUS.getFieldName());
        add(Field.CHANNEL.getFieldName());
        add(Field.PAYLOAD.getFieldName());
        add(Field.STICKY.getFieldName());
        add(Field.EXPIRE.getFieldName());
    }};

    private void checkUpstreamExtraFields(Map<String, Object> data) {
        for (String field : data.keySet()) {
            if (!UPSTREAM_FIELDS.contains(field)) {
                throw new InvalidRequestException("Invalid field in upstream message: " + field);
            }
        }
    }

    /**
     * @return a time-based, lexicographically comparable message ID.
     */
    private static String generateMessageId(Date date) {
        return DateTimeUtils.ISO8601.get().format(date) + "-" + RandomUtils.randomString(10);
    }

    private String extractFieldValueAsJsonString(Field field, Map<String,Object> data) throws BackplaneServerException {
        try {
            Object value = data.get(field.getFieldName());
            return value == null ? null : (new ObjectMapper()).writeValueAsString(value);
        } catch (IOException e) {
            String errMsg = "Error serializing message payload: " + e.getMessage();
            logger.error(errMsg);
            throw new BackplaneServerException(errMsg, e);
        }
    }

}

File: src/main/scala/com/janrain/backplane/server2/model/BackplaneMessage.scala
package com.janrain.backplane.server2.model

import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}
import com.janrain.backplane2.server.Scope
import com.janrain.backplane2.server.Scope.ScopeType._
import com.janrain.servlet.InvalidRequestException
import com.janrain.backplane.common.{BackplaneServerException, DateTimeUtils}
import java.util.Date
import com.janrain.util.{Loggable, RandomUtils}
import org.codehaus.jackson.map.ObjectMapper
import java.io.IOException
import scala.collection.JavaConversions._
import com.janrain.backplane2.server.Scope.ScopeType

/**
 * @author Johnny Bufu
 */
class BackplaneMessage(data: Map[String,String]) extends Message(data, BackplaneMessageFields.values) {

  def this(clientSourceUrl: String, defaultExpireSeconds: Int, maxExpireSeconds: Int, upstreamData: java.util.Map[String,AnyRef]) =
    this(
      BackplaneMessage.processExpire(
        BackplaneMessage.parseUpstreamData(upstreamData.toMap), defaultExpireSeconds, maxExpireSeconds )
      ++
      Map(
        BackplaneMessageFields.ID.name -> BackplaneMessage.generateMessageId(new Date),
        BackplaneMessageFields.SOURCE.name -> clientSourceUrl
      ))

  def idField = BackplaneMessageFields.ID

  def asFrame(serverDomain: String, includePayload: Boolean): Map[String, AnyRef] =
    BackplaneMessageFields.values
      .map(field => field.name -> field.frameOutput(id, serverDomain, includePayload, get(field))).toMap
      .filter { case (k,v) => v.isDefined } .mapValues(_.get)

  def bus: String = get(BackplaneMessageFields.BUS)
    .getOrElse(throw new IllegalStateException("bus field missing from bp2 message, should have failed validation"))

  def channel: String = get(BackplaneMessageFields.CHANNEL)
    .getOrElse(throw new IllegalStateException("channel field missing from bp2 message, should have failed validation"))

  def expiration: String = get(BackplaneMessageFields.EXPIRE)
    .getOrElse(throw new IllegalStateException("expire field missing from bp2 message, should have failed validation"))
}

object BackplaneMessage {

  private final val MESSAGE_ID_RANDOM_LENGTH = 10
  private final val MESSAGE_ID_LEGACY_RANDOM_LENGTH = 10

  private final val UPSTREAM_FIELDS = BackplaneMessageFields.values.filter(_.isUpstream).toList
  private final val UPSTREAM_FIELD_NAMES = UPSTREAM_FIELDS.map(_.name)

  val scopeKeys = BackplaneMessageFields.values.filter(_.scopeType != ScopeType.NONE).map(sc => sc.name -> sc).toMap

  def generateMessageId(date: Date) =
    DateTimeUtils.ISO8601.get.format(date) + "-" + RandomUtils.randomString(MESSAGE_ID_RANDOM_LENGTH)

  def legacyIdLength: Int = 25 + MESSAGE_ID_LEGACY_RANDOM_LENGTH // ISO8601 length + "-" + random_length

  private def parseUpstreamData(upstreamData: Map[String,AnyRef]): Map[String,String] = {
    BackplaneMessage.checkUpstreamFields(upstreamData)
    BackplaneMessage.UPSTREAM_FIELDS
      .map(field => field.name -> field.parseUpstreamValue(upstreamData)).toMap
      .filter { case (k,v) => v.isDefined } .mapValues(_.get)
  }

  private def checkUpstreamFields(upstreamData: Map[String,AnyRef]) {
    upstreamData.keySet.foreach( field =>
      if (! UPSTREAM_FIELD_NAMES.contains(field))
        throw new InvalidRequestException("Invalid field in upstream message: " + field))
  }

  private def processExpire(data: Map[String,String], defaultExpSeconds: Int, stickyExpSeconds: Int): Map[String,String] = {
    val sticky = data.get(BackplaneMessageFields.STICKY.name).exists(_.equalsIgnoreCase(true.toString))
    data.map( kv =>
      if (kv._1 == BackplaneMessageFields.EXPIRE.name) kv._1 -> DateTimeUtils.processExpireTime(sticky, kv._2, defaultExpSeconds, stickyExpSeconds)
      else kv
    )
  }
}

object BackplaneMessageFields extends MessageFieldEnum with Loggable {

  type BackplaneMessageField = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
    def isUpstream = true
    def parseUpstreamValue(upstreamData: Map[String,AnyRef]): Option[String] = upstreamData.get(name).map(_.toString)
    def scopeType: Scope.ScopeType
    def frameOutput(msgId: String, serverDomain: String, includePayload: Boolean, fieldValue: Option[String]): Option[AnyRef] = fieldValue
  }

  val ID = new BackplaneMessageField { def name = "id"; def scopeType = NONE
    override def isUpstream = false
    override def frameOutput(msgId: String, serverDomain: String, includePayload: Boolean, fieldValue: Option[String]) = None
    override def validateLong(fieldValue: Option[String]) {
      super.validateLong(fieldValue)
      fieldValue.foreach(Message.dateFromId)
    }
  }

  val CHANNEL = new BackplaneMessageField { def name = "channel"; def scopeType = FILTER }

  val BUS = new BackplaneMessageField { def name = "bus"; def scopeType = AUTHZ_REQ }

  val STICKY = new BackplaneMessageField { def name = "sticky"; def scopeType = FILTER
    override def required = false
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateBoolean(fieldValue)
    }
    override def parseUpstreamValue(upstreamData: Map[String, AnyRef]) =
      upstreamData.get(name).map(_.toString).orElse(Option(false.toString))
  }

  val EXPIRE = new BackplaneMessageField { def name = "expire"; def scopeType = NONE
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateInternetDate(fieldValue)
    }
  }

  val SOURCE = new BackplaneMessageField { def name = "source"; def scopeType = FILTER
    override def isUpstream = false
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateUrl(fieldValue)
    }
  }

  val TYPE = new BackplaneMessageField { def name = "type"; def scopeType = FILTER }

  val MESSAGE_URL = new BackplaneMessageField { def name = "message_url"; def scopeType = FILTER
    override def required = false
    override def isUpstream = false
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateUrl(fieldValue)
    }
    override def frameOutput(msgId: String, serverDomain: String, includePayload: Boolean, fieldValue: Option[String]) =
      Some("https://" + serverDomain + "/v2/message/" + msgId)
  }

  val PAYLOAD = new BackplaneMessageField { def name = "payload"; def scopeType = NONE
    override def parseUpstreamValue(upstreamData: Map[String, AnyRef]) = {
      try {
        upstreamData.get(name).map(new ObjectMapper().writeValueAsString(_))
      } catch {
        case e: IOException => {
          val errMsg = "Error serializing message payload: " + e.getMessage
          logger.error(errMsg)
          throw new BackplaneServerException(errMsg, e)
        }
      }
    }
    override def frameOutput(msgId: String, serverDomain: String, includePayload: Boolean, fieldValue: Option[String]) = {
      if (includePayload) {
        try {
          Some(new ObjectMapper().readValue(fieldValue.getOrElse(null), classOf[AnyRef]) ) // un-quote the value
        } catch {
          case e: IOException => {
            val errMsg = "Error deserializing message payload: " + e.getMessage
            logger.error(errMsg)
            throw new BackplaneServerException(errMsg, e)
          }
        }
      } else {
        None
      }
    }
  }

}
